agent/logging.go

```diff
@@ -123,97 +116,85 @@ func (a *Agent) makeLoggerProducer(ctx context.Context, tc *taskContext, c *mode
-		}
-		splunkToken, err := tc.expansions.ExpandString(agentConfig.SplunkToken)
-		if err != nil {
-			grip.Error(errors.Wrap(err, "error expanding splunk token"))
-		}
-		config.Agent = append(config.Agent, client.LogOpts{
-			LogkeeperURL:      a.opts.LogkeeperURL,
-			LogkeeperBuilder:  tc.taskModel.Id,
-			LogkeeperBuildNum: tc.taskModel.Execution,
-			Sender:            agentConfig.Type,
-			SplunkServerURL:   splunkServer,
-			SplunkToken:       splunkToken,
-			Filepath:          filepath.Join(path, agentLogFileName),
-		})
+		a.prepSingleLogger(tc, agentConfig, &config.Agent, logDir, agentLogFileName)
 	}
 	for _, systemConfig := range c.System {
-		splunkServer, err := tc.expansions.ExpandString(systemConfig.SplunkServer)
-		if err != nil {
-			grip.Error(errors.Wrap(err, "error expanding splunk server"))
-		}
-		splunkToken, err := tc.expansions.ExpandString(systemConfig.SplunkToken)
-		if err != nil {
-			grip.Error(errors.Wrap(err, "error expanding splunk token"))
-		}
-		config.System = append(config.System, client.LogOpts{
-			LogkeeperURL:      a.opts.LogkeeperURL,
-			LogkeeperBuilder:  tc.taskModel.Id,
-			LogkeeperBuildNum: tc.taskModel.Execution,
-			Sender:            systemConfig.Type,
-			SplunkServerURL:   splunkServer,
-			SplunkToken:       splunkToken,
-			Filepath:          filepath.Join(path, systemLogFileName),
-		})
+		a.prepSingleLogger(tc, systemConfig, &config.System, logDir, systemLogFileName)
 	}
 	for _, taskConfig := range c.Task {
-		splunkServer, err := tc.expansions.ExpandString(taskConfig.SplunkServer)
-		if err != nil {
-			grip.Error(errors.Wrap(err, "error expanding splunk server"))
-		}
-		splunkToken, err := tc.expansions.ExpandString(taskConfig.SplunkToken)
-		if err != nil {
-			grip.Error(errors.Wrap(err, "error expanding splunk token"))
-		}
-		config.Task = append(config.Task, client.LogOpts{
-			LogkeeperURL:      a.opts.LogkeeperURL,
-			LogkeeperBuilder:  tc.taskModel.Id,
-			LogkeeperBuildNum: tc.taskModel.Execution,
-			Sender:            taskConfig.Type,
-			SplunkServerURL:   splunkServer,
-			SplunkToken:       splunkToken,
-			Filepath:          filepath.Join(path, taskLogFileName),
-		})
+		a.prepSingleLogger(tc, taskConfig, &config.Task, logDir, taskLogFileName)
 	}
 
 	return config
 }
 
+func (a *Agent) prepSingleLogger(tc *taskContext, in model.LogOpts, out *[]client.LogOpts, logDir, fileName string) {
+	splunkServer, err := tc.expansions.ExpandString(in.SplunkServer)
+	if err != nil {
+		grip.Error(errors.Wrap(err, "error expanding splunk server"))
+	}
+	splunkToken, err := tc.expansions.ExpandString(in.SplunkToken)
+	if err != nil {
+		grip.Error(errors.Wrap(err, "error expanding splunk token"))
+	}
+	if in.LogDirectory != "" {
+		grip.Error(errors.Wrap(os.MkdirAll(in.LogDirectory, os.ModeDir|os.ModePerm), "error making log directory"))
+		logDir = in.LogDirectory
+	}
+	tc.logDirectories = append(tc.logDirectories, logDir)
+	*out = append(*out, client.LogOpts{
```

Comment:
can we return this rather than having a pointer to a slice?

Reply:
changed